sequence_processing_pipeline/Commands.py

```diff
@@ -0,0 +1,151 @@
+                    f2.write("Hello")
+            else:
+                print(".")
+
+
+def split_similar_size_bins(data_location_path, max_file_list_size_in_gb,
+                            batch_prefix):
+    '''
+
+    :param data_location_path: Path to the ConvertJob directory.
+    :param max_file_list_size_in_gb: Upper threshold for file-size.
+    :param batch_prefix: Path + file-name prefix for output-files.
+    :return: The number of output-files created.
+    '''
+
+    # SPP root paths are of the form:
+    # ../72dc6080-c453-418e-8a47-1ccd6d646a30/ConvertJob, and contain only
+    # directories named after projects e.g:
+    # 72dc6080-c453-418e-8a47-1ccd6d646a30/ConvertJob/Tell_Seq_15196.
+    # Each of these directories contain R1 and R2 fastq files. Hence, path
+    # is now the following:
+    fastq_paths = glob.glob(data_location_path + '*/*.fastq.gz')
+    fastq_paths.sort()
+
+    # output_base_count is the number of fastq files in data_location_path.
+    # fastq_paths should contain a list of R1 and R2 fastq files only.
+    output_base_count = len(fastq_paths)
+
+    files = iter(fastq_paths)
+
+    # convert from GB and halve as we sum R1
+    max_size = (int(max_file_list_size_in_gb) * (2 ** 30) / 2)
+
+    r1 = iter(files)
+    r2 = iter(files)
+
+    split_offset = 0
+
+    # ensure we are max-sized to start.
+    current_size = max_size * 10
+    fp = None
+
+    for a, b in zip(r1, r2):
+        # a: /some_path/13722/115207/TCGA-05-4395-01A-01D-1203_110913_SN208_
+        #  0238_BC008YACXX_s_8_rg.sorted.filtered.R1.trimmed.fastq.gz
+        # b: /some_path/13722/115207/TCGA-05-4395-01A-01D-1203_110913_SN208_
+        #  0238_BC008YACXX_s_8_rg.sorted.filtered.R2.trimmed.fastq.gz
+        if not 'R1' in a:
+            raise ValueError(f"{a} is not a forward-read fastq file.")
+
+        if not 'R2' in b:
+            raise ValueError(f"{b} is not a reverse-read fastq file.")
+
+        # assert that paths a and b are identical up to where R1 and R2 are
+        # present.
+        if not a[:a.find('.R1')] == b[:b.find('.R2')]:
+            raise ValueError(f"{a} and {b} are not paired fastq files.")
+
+        r1_size = os.stat(a).st_size
+
+        # output_base is the path up to the beginning of the filename.
+        # we could do this in theory with os.path.split()[0] e.g.:
+        # /panfs/dtmcdonald/human-depletion/t2t-only/13722/115207
+        # /panfs/dtmcdonald/human-depletion/t2t-only/13722/115212
+        output_base = split(a)
+
+        if current_size + r1_size > max_size:
+            if fp is not None:
+                fp.close()
+            split_offset += 1
+            current_size = r1_size
+            fp = open(batch_prefix + '-%d' % split_offset, 'w')
+        else:
+            current_size += r1_size
+
+        fp.write("%s\t%s\t%s\n" % (a, b, output_base))
+    fp.close()
+
+    return split_offset
+
+
+def demux_inline(argv1, argv2, argv3, argv4):
```

Comment:
This strategy is designed to read over the multiplexed file multiple times. We now use a strategy that reads once, and parallelizes higher in processing. I'm not sure what is better for Qiita

Reply:
@antgonza I'm thinking differences in memory usage might be a meaningful way to judge one over the other. I'll fix this PR and then read over the new version and see if I can't add it as an alternate. Decoupling this will make it easier to test something in Qiita-RC earlier.

Reply:
Superseded by Daniel's demux(id_map, fp, out_d, encoded, threads) function.